### PR TITLE
fix: sweep legacy target-resolved File identities with positive ownership attribution

### LINKED ISSUE #1156: Incremental --update-graph does not reconcile legacy target-resolved File identities
Follow-up from Greptile's review of #1155.

## Context
#1155 keys `File` nodes on the link path (leaf not dereferenced) so a deleted symlink is prunable. That is correct for all newly written nodes.

## Residual gap (narrow)
On an **incremental** `--update-graph` run over a graph originally built by an **older** version, a legacy `File` node that was keyed on a symlink's *dereferenced target* may persist:
- In-root symlinks self-heal: the real file MERGEs its own key, the link gets its new key, both paths exist so prune keeps both. No stale data.
- The genuine residual is an **external** symlink recorded in a **pre-GHSA-85gg** graph: its stored `absolute_path` is outside the repo, so the prune containment gate skips it and it is never swept.

## Why this does not block #1155
- The standard upgrade path is `--clean` (full rebuild via `clean_database()`, cli.py:300), which wipes the graph — no legacy data survives.
- Post-0.0.589, external symlinks are skipped at ingest, so new graphs never create these nodes.
- The old code could not delete or prune these nodes either, so #1155 is strictly better, not a regression.

## Possible fix
During `_prune_orphan_nodes`, also drop a `File` node whose stored `absolute_path` disagrees with the identity derivable from its relative `path` (`cached_file_identity_posix(repo_path / path)`), i.e. a legacy target-resolved record. Needs careful handling of symlinked-ancestor edge cases and its own RED-verified test.

### PR DESCRIPTION
**Pull request by vitali87**
Closes #1156.

A pre-GHSA-85gg graph can hold a `File` node keyed on an external symlink's dereferenced target. Its `absolute_path` sits outside the repository, so the prune containment gate skips it forever; `--clean` wipes it, but an incremental `--update-graph` over a legacy graph never does.

## What

During `_prune_orphan_nodes`, an out-of-repo File key becomes a sweep candidate only when it disagrees with the identity derivable from its relative path (`cached_file_identity_posix(repo_path / path)`). A key that agrees is a file under a symlinked ancestor directory, which is legitimate under the current scheme (parent resolved, leaf kept) and is never a candidate, so the symlinked-ancestor edge case the issue flagged costs nothing.

The delicate part is that File nodes MERGE globally on `absolute_path`: the stale key can be the very node another project legitimately owns (its repo contains the old link's target), and the relative path can also collide with a sibling root (issue #897). Candidates are therefore deleted only under **positive ownership attribution** via a per-candidate `CONTAINS`-containers query: every container must belong to this project (the Project node by name, Folders by containment under the repo root) and at least one must exist. Foreign containers veto; missing or unreadable containers veto, because absence of evidence is not sole ownership. The pre-existing #897 sibling-root tests were exactly what caught the naive version of this sweep deleting a sibling's node, and they now pass against the attribution rule.

## Tests

RED-first: the external legacy key is swept when contained solely by this project's folders; a foreign Folder or a foreign Project container spares the key (parametrised); a key with no visible containers is spared; and a real on-disk directory-symlink fixture pins that a current-scheme symlinked-ancestor identity is never even a candidate (no containers query issued). The 148-test prune/incremental/symlink battery passes.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved cleanup of legacy file records created from symlink-resolved paths.
  - Removes stale records only when ownership is clearly limited to the current project.
  - Preserves records shared with other projects or where ownership is uncertain.
  - Prevents valid file identities created through directory symlinks from being removed.
  - Handles missing or inaccessible ownership information safely.

- **Chores**
  - Added clearer logging for legacy file cleanup operations.

### CODE REVIEW

File: codebase_rag/constants/graph.py [hunk 1]
```diff
@@ -399,6 +399,14 @@ class AuditCheck(StrEnum):
 CYPHER_ALL_FILE_PATHS = (
     "MATCH (f:File) RETURN f.path AS path, f.absolute_path AS absolute_path"
 )
+# Containers of one File key, for legacy-identity sweep attribution: File
+# nodes MERGE globally on absolute_path, so a key can be shared with another
+# project and must not be deleted from under it (issue #1156).
+CYPHER_FILE_CONTAINERS = (
+    "MATCH (p)-[:CONTAINS]->(f:File {absolute_path: $path}) "
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/constants/graph.py]: Fixed in 43cb6165, and this was a genuine defect: the query matched CONTAINS while File parents link via CONTAINS_FILE, so the sweep would have found no owners and spared everything behind the positive-attribution rule (inert, though never destructive). Corrected to CONTAINS_FILE, with a schema-drift test asserting the query names the same relationship constant structure_processor writes.

File: codebase_rag/graph_updater.py [hunk 2]
```diff
@@ -2279,6 +2294,55 @@ def _prune_orphan_nodes(self) -> None:
...
+    def _file_key_owned_only_by_this_project(self, abs_path: str) -> bool:
+        """Positive attribution: every container is this project's, and at
+        least one exists. A sibling project's node can share both the key
+        and the relative path (issue #897), and a node whose containers are
+        missing or unreadable offers no evidence of sole ownership, so both
+        veto the sweep.
+        """
+        rows = self.ingestor.fetch_all(
+            cs.CYPHER_FILE_CONTAINERS, {cs.KEY_PATH: abs_path}
+        )
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/graph_updater.py]: Fixed in 43cb6165: the per-candidate ownership read is wrapped; a failure logs the same prune-read warning, spares the key (unreadable ownership is unknown ownership), and never escapes the prune mid-update. Regression test raises from the containers query and asserts no delete and no exception.

File: codebase_rag/graph_updater.py [hunk 3]
```diff
@@ -2279,6 +2294,55 @@ def _prune_orphan_nodes(self) -> None:
...
+                owned = True
+                continue
+            if isinstance(container_abs, str) and container_abs:
+                if container_abs == repo_abs or container_abs.startswith(
+                    repo_abs + "/"
+                ):
+                    owned = True
+                else:
+                    return False
+        return owned
```
[the remark this answers is not in the document]
Comment (vitali87) [reply codebase_rag/graph_updater.py]: Fixed in 43cb6165: a container row with no usable identity now vetoes outright rather than being skipped, so one local owner plus one unreadable row spares the key. Covered by a dedicated test with exactly that pair.
